zerocopy-derive/src/enums.rs

```diff
@@ -0,0 +1,491 @@
+
+                impl #variant_struct_ident #impl_generics #where_clause {
+                    fn is_bit_valid<A>(this: ::zerocopy::Maybe<'_, Self, A>) -> bool
+                    where
+                        A: ::zerocopy::pointer::invariant::Aliasing
+                            + ::zerocopy::pointer::invariant::AtLeast<::zerocopy::pointer::invariant::Shared>,
+                    { true }
+                }
+            }
+        }
+    }
+}
+
+fn derive_is_bit_valid_c(data: &DataEnum, generics: &Generics, idents: &RawIdents) -> TokenStream {
+    let discriminant_type = generate_discriminant_type(data, &EnumRepr::C, &idents.discriminant);
+    let discriminant_consts = generate_discriminant_consts(data, idents);
+
+    let non_unit_variants =
+        data.variants.iter().filter(|variant| !matches!(variant.fields, Fields::Unit));
+    let variant_structs = non_unit_variants
+        .clone()
+        .map(|variant| generate_variant_struct(variant, generics, idents, false));
+
+    let raw_ident = &idents.r#enum;
+    let discriminant_ident = &idents.discriminant;
+    let fields_ident = &idents.fields;
+
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let variant_idents = non_unit_variants.clone().map(|v| &v.ident);
+    let variant_struct_idents = non_unit_variants.clone().map(|v| idents.variant_struct(&v.ident));
+
+    let match_arms = data.variants.iter().map(|variant| {
+        let value_ident = idents.discriminant_const(&variant.ident);
+
+        if matches!(variant.fields, Fields::Unit) {
+            quote! {
+                #value_ident => true
+            }
+        } else {
+            let variant_struct = idents.variant_struct(&variant.ident);
+
+            quote! {
+                #value_ident => {
+                    // SAFETY:
+                    // - `cast_unsized`:
+                    //   - The variant struct has the same size as the enum
+                    //   - The returned pointer is cast from `p`, and so has the
+                    //     same provenance as `p`.
+                    //   - We checked that the discriminant of the enum matched
+                    //     the value for this variant, so this cast preserves
+                    //     types and locations of all fields. Therefore, any
+                    //     `UnsafeCell`s will have the same location as in the
+                    //     original type.
+                    // - `assume_initialized`: `cast_unsized` removes the
+                    //   initialization invariant from `p`, so we re-assert that
+                    //   all of the bytes are initialized.
+                    let candidate = unsafe {
+                        fields
+                            .cast_unsized(|p: *mut #fields_ident| p as *mut #variant_struct)
+                            .assume_initialized()
+                    };
+                    #variant_struct::is_bit_valid(candidate)
```

Comment:
Here and elsewhere: should this be fully-qualified?

Reply:
For variant structs, no. These are inherent impls that have the same name as `TryFromBytes`'s methods. To keep the generated code down as much as possible, the variant structs don't actually implement `TryFromBytes`. Still, these should probably be `<#variant_struct>::is_bit_valid` so I changed that.